packages/flame/lib/src/effects2/controllers/effect_controller.dart

```diff
@@ -0,0 +1,170 @@
+/// variable [progress], which will grow from 0.0 to 1.0. The value of 0
+/// corresponds to the beginning of an animation, and the value of 1.0 is
+/// the end of the animation.
+///
+/// The [progress] variable can best be thought of as a "logical time". For
+/// example, if you want to animate a certain property from value A to value B,
+/// then you can use [progress] to linearly interpolate between these two
+/// extremes and obtain variable `x = A*(1 - progress) + B*progress`.
+///
+/// The exact behavior of [progress] is determined by subclasses, but the
+/// following general considerations apply:
+///   - the progress can also go in negative direction (i.e. from 1 to 0);
+///   - the progress may oscillate, going from 0 to 1, then back to 0, etc;
+///   - the progress may change over a finite or infinite period of time;
+///   - the value of 0 corresponds to the logical start of an animation;
+///   - the value of 1 is either the end or the "peak" of an animation;
+///   - the progress may briefly attain values outside of [0; 1] range (for
+///     example if a "bouncy" easing curve is applied).
+///
+/// An [EffectController] can be made to run forward in time (`advance()`), or
+/// backward in time (`recede()`).
+///
+/// Unlike the `dart.ui.AnimationController`, this class does not use a `Ticker`
+/// to keep track of time. Instead, it must be pushed through time manually, by
+/// calling the `update()` method within the game loop.
+abstract class EffectController {
+  /// Factory function for producing common [EffectController]s.
+  ///
+  /// In the simplest case, when only `duration` is provided, this will return
+  /// a [LinearEffectController] that grows linearly from 0 to 1 over the period
+  /// of that duration.
+  ///
+  /// More generally, the produced effect controller allows to add a delay
+  /// before the beginning of the animation, to animate both forward and in
+  /// reverse, to iterate several times (or infinitely), to apply an arbitrary
+  /// [curve] making the effect progression non-linear, etc.
+  ///
+  /// In the most general case, the animation proceeds through the following
+  /// steps:
+  ///   1. wait for [startDelay] seconds,
+  ///   2. repeat the following steps [repeatCount] times (or [infinite]ly):
+  ///       a. progress from 0 to 1 over the [duration] seconds,
+  ///       b. wait for [atMaxDuration] seconds,
+  ///       c. progress from 1 to 0 over the [reverseDuration] seconds,
+  ///       d. wait for [atMinDuration] seconds.
+  ///
+  /// Setting parameter [alternate] to true is another way to create a
+  /// controller whose [reverseDuration] is the same as the forward [duration].
+  ///
+  /// If the animation is finite and there are no "backward" or "atMin" stages
+  /// then the animation will complete at `progress == 1`, otherwise it will
+  /// complete at `progress == 0`.
+  factory EffectController({
+    required double duration,
+    Curve? curve,
```

Comment:
Maybe this should default to `linear` so that it is easier to see in the IDE what the default is, the code should become slightly simpler too.